baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferInputStreamTransferTo.java

```diff
@@ -0,0 +1,111 @@
+                + "where as on JDK < 21 and when using Guava `ByteStreams.copy` 8KiB byte chunk copies are used. "
+                + "\n\nReferences:\n\n"
+                + "  * https://github.com/palantir/hadoop-crypto/pull/586\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8299336\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8067661\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8265891\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8273038\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8279283\n"
+                + "  * https://bugs.openjdk.org/browse/JDK-8296431\n")
+public final class PreferInputStreamTransferTo extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final String ERROR_MESSAGE = "Prefer InputStream.transferTo(OutputStream)";
+
+    public static final String INPUT_STREAM = "java.io.InputStream";
+    private static final Matcher<Tree> INPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(INPUT_STREAM);
+    public static final String OUTPUT_STREAM = "java.io.OutputStream";
+    private static final Matcher<Tree> OUTPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(OUTPUT_STREAM);
+
+    private static final Matcher<ExpressionTree> GUAVA_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("com.google.common.io.ByteStreams")
+            .namedAnyOf("copy")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+    private static final Matcher<ExpressionTree> APACHE_COMMONS_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("org.apache.commons.io.IOUtils")
+            .namedAnyOf("copy", "copyLarge")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+    private static final Matcher<ExpressionTree> AWS_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("com.amazonaws.util.IOUtils")
+            .namedAnyOf("copy")
+            .withParameters(INPUT_STREAM, OUTPUT_STREAM);
+
+    private static final Matcher<ExpressionTree> METHOD_MATCHER = Matchers.anyOf(
+            GUAVA_BYTE_STREAM_COPY_MATCHER, APACHE_COMMONS_BYTE_STREAM_COPY_MATCHER, AWS_BYTE_STREAM_COPY_MATCHER);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (METHOD_MATCHER.matches(tree, state)) {
+            List<? extends ExpressionTree> args = tree.getArguments();
+            if (args.size() != 2) {
+                return Description.NO_MATCH;
+            }
+
+            ExpressionTree maybeInputStreamArg = args.get(0);
+            ExpressionTree maybeOutputStreamArg = args.get(1);
+            if (INPUT_STREAM_MATCHER.matches(maybeInputStreamArg, state)
+                    && OUTPUT_STREAM_MATCHER.matches(maybeOutputStreamArg, state)) {
```

Comment:
The signatures that we match already require these to be an input stream and outputstream, but I suppose future methods we discover may be broader and I don’t mind being especially defensive in this sort of check.
It might be worthwhile to add some defensiveness against updating a transferTo implementation to call itself. We could exclude invocations using ‘this’ on the lhs, with the assumption that inputstream subtypes may require a closer look.

Reply:
Yeah, might be worth simplifying the arg checks now.

Reply:
> It might be worthwhile to add some defensiveness against updating a transferTo implementation to call itself. We could exclude invocations using ‘this’ on the lhs, with the assumption that inputstream subtypes may require a closer look.

I think this would likely still be a valid refactor. I could imagine some  custom InputStream that had a `writeTo(OutputStream out)` method that used say `ByteStreams.copy(this, out)`

Reply:
The case that I have in mind here is

```diff
@Override
public long transferTo(OutputStream other) throws IOException {
-    ByteStreams.copy(this, other);
+    this.transferTo(other); // Infinite recursion!
}

Reply:
I think either way you prefer is reasonable. If I were writing this, I would probably leave as-is for `this` on the lhs since I expect more edge cases while implementing an InputStream compared to consuming one. My philosophy with this sort of check is that we should catch the common cases where we have the most confidence in our recommendations, and use those to teach folks the intricacies which they can apply to more ambiguous cases. But again, it's a subtle difference, and I'm happy to move forward with either.